### ISSUE TITLE
pp ({"a" => nil}.compact) leads to a compiler bug

### ISSUE THREAD

**Issue by vladfaust** [crystal-lang/crystal#7019]
See https://carc.in/#/r/5f1o

**Comment by r00ster91** [comment]
Reduced:
```crystal
hash = {nil => nil}
hash.each.with_object({} of Nil => typeof(hash.first_value.not_nil!)).to_s
```
Without `not_nil!` or `to_s` the bug doesn't happen.

**Comment by bew** [comment]
More reduced: https://carc.in/#/r/5f70
```cr
puts Hash(Nil, typeof(nil.not_nil!)).new
```

**Comment by RX14** [comment]
Fixed in 0.27.0.

**Comment by RX14** [comment]
Most likely https://github.com/crystal-lang/crystal/pull/6659.
